src/layers/mapbox-utils.js

```diff
@@ -139,37 +139,68 @@ function updateSourceData(map, sourceId, data) {
  * @param properties [{label: {fieldIdx}]
  * @returns {{type: string, properties: {}, features: {type: string, properties: {}, geometry: {type: string, coordinates: *[]}}[]}}
  */
-export function geojsonFromPoints(
+export function geoJsonFromData(
   allData = [],
   filteredIndex = [],
-  columns = {},
-  properties = []
+  getGeometry,
+  getProperties = (d, i) => {}
 ) {
+
   const geojson = {
     type: 'FeatureCollection',
     features: []
   };
 
   for (let i = 0; i < filteredIndex.length; i++) {
-    const point = allData[filteredIndex[i]];
-    geojson.features.push({
-      type: 'Feature',
-      properties: properties.reduce(
-        (final, property) => ({
-          ...final,
-          [property.name]: point[property.tableFieldIndex - 1]
-        }),
-        {}
-      ),
-      geometry: {
-        type: 'Point',
-        coordinates: [
-          columns.lng ? point[columns.lng.fieldIdx] : null, // lng
-          columns.lat ? point[columns.lat.fieldIdx] : null // lat
-        ]
-      }
-    });
+    const index = filteredIndex[i];
+    const point = allData[index];
+    const geometry = getGeometry(point);
+
+    if (geometry) {
+      geojson.features.push({
+        type: 'Feature',
+        properties: {
+          index,
+          ...getProperties(point, index)
+        },
+        geometry
+      });
+    }
   }
 
   return geojson;
 }
+
+export const prefixGpuField = (name) => `gpu:${name}`
+
+export function gpuFilterToMapboxFilter(gpuFilter) {
+  const {
+    filterRange,
+    filterValueUpdateTriggers
+  } = gpuFilter;
```

Comment:
this is related to the previous comment, gpuFilter may be null

Reply:
done